### ISSUE TITLE
Support methods like NCE, NEG, Hierarchical SoftMax, Blackout or IS

### ISSUE THREAD

**Issue by michellemay** [intel-analytics/BigDL#2236]
When training a classifier with large number of classes, performance sink. This is expected when using regular (log)SoftMax methods to compute the loss since it needs to normalize current class score with the sum of all other classes score.  

I think this would be helpful to have approximate methods like Hierarchical SoftMax, NCE, NEG, IS to speedup training.

A paper comparing different methods for approximate normalization over all classes:
http://web4.cs.ucl.ac.uk/staff/D.Barber/publications/AISTATS2017.pdf
